libbpf-tools/funcinterval.c

```diff
@@ -0,0 +1,391 @@
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		lookup_key = next_key;
+	}
+
+	lookup_key = -1;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+
+	fd = bpf_map__fd(obj->maps.start);
+	lookup_key = -1;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup start : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+
+	return 0;
+}
+
+static int split_pattern(const char *raw_pattern, enum TRACE_TYPE *type,
+			 const char **library, const char **pattern)
+{
+	const char *string1, *string2, *string3;
+	char *raw_pattern_tmp = strdup(raw_pattern);
+
+	string1 = strsep(&raw_pattern_tmp, ":");
+	if (!raw_pattern_tmp) {
+		/* Not found ':', return raw_pattern */
+		*type = KPROBE;
+		*pattern = raw_pattern;
+		return 0;
+	}
+
+	string2 = strsep(&raw_pattern_tmp, ":");
+	if (!raw_pattern_tmp) {
+		/* One ':' found, return library */
+		*type = UPROBE;
+		*library = string1;
+		*pattern = string2;
+		return 0;
+	}
+
+	string3 = strsep(&raw_pattern_tmp, ":");
+	if (strlen(string1) != 1)
+		return -EINVAL;
+	else if (string1[0] == 't')
+		*type = TRACEPOINT;
+	else if (string1[0] == 'p') {
+		if (strlen(string2) == 0)
+			*type = KPROBE;
+		else
+			*type = UPROBE;
+	} else
+		return -EINVAL;
+
+	if (*type != KPROBE)
+		*library = string2;
+	*pattern = string3;
+
+	return 0;
+}
+
+
+static int attach_uprobe(struct funcinterval_bpf *obj, const char *binary,
+			 const char *function)
+{
+	int pid = env.pid;
+	char bin_path[PATH_MAX];
+	off_t func_off;
+
+	if (pid == 0)
+		pid = getpid();
+
+	if (resolve_binary_path(binary, pid, bin_path, sizeof(bin_path)))
+		return 1;
+
+	func_off = get_elf_func_offset(bin_path, function);
+	if (func_off < 0) {
+		fprintf(stderr, "Could not find %s in %s\n", function, bin_path);
+		return 1;
+	}
+
+	obj->links.function_uprobe_entry =
+		bpf_program__attach_uprobe(obj->progs.function_uprobe_entry,
+					   false, pid ?: -1, bin_path, func_off);
+	if (!obj->links.function_uprobe_entry) {
+		fprintf(stderr, "Failed to attach uprobe: %d\n", -errno);
+		return 1;
+	}
+
+	return 0;
+}
+
+static int attach_tracepoint(struct funcinterval_bpf *obj, const char *library,
+			     const char *pattern)
+{
+	obj->links.tracepoint_entry =
+		bpf_program__attach_tracepoint(obj->progs.tracepoint_entry,
+					       library, pattern);
+	if (!obj->links.tracepoint_entry) {
+		fprintf(stderr, "Failed to attach t:%s:%s\n", library, pattern);
+		return -errno;
+	}
+	return 0;
+}
+
+int main(int argc, char *argv[])
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct funcinterval_bpf *obj;
```

Comment:
Maybe you can add `ensure_core_btf` and  `cleanup_core_btf` like
 https://github.com/iovisor/bcc/blob/6acb86effa7a6e8029b68eccb805dd1ee60ecc5a/libbpf-tools/funclatency.c#L349

Reply:
I have modified and submitted it

Reply:
how about squashing this into the previous commit?